# This is Jeopardy!

I have been provided an unmodified csv file containing data about the game show _Jeopardy!_ in a file named `jeopardy.csv`.

In [1]:
import pandas as pd
pd.set_option('display.max_colwidth', None) # Displays the full column name when printing
all_data = pd.read_csv('jeopardy.csv')
print(all_data.columns)

Index(['Show Number', ' Air Date', ' Round', ' Category', ' Value',
       ' Question', ' Answer'],
      dtype='object')


The column names are not named cleanly, so I choose to fix this now.

In [2]:
all_data.columns=['Show Number', 'Air Date', 'Round', 'Category', 'Value', 'Question', 'Answer']
print(all_data.columns)

Index(['Show Number', 'Air Date', 'Round', 'Category', 'Value', 'Question',
       'Answer'],
      dtype='object')


In [3]:
all_data

Show Number    Air Date             Round  \
0              4680  2004-12-31         Jeopardy!   
1              4680  2004-12-31         Jeopardy!   
2              4680  2004-12-31         Jeopardy!   
3              4680  2004-12-31         Jeopardy!   
4              4680  2004-12-31         Jeopardy!   
...             ...         ...               ...   
216925         4999  2006-05-11  Double Jeopardy!   
216926         4999  2006-05-11  Double Jeopardy!   
216927         4999  2006-05-11  Double Jeopardy!   
216928         4999  2006-05-11  Double Jeopardy!   
216929         4999  2006-05-11   Final Jeopardy!   

                               Category  Value  \
0                               HISTORY   $200   
1       ESPN's TOP 10 ALL-TIME ATHLETES   $200   
2           EVERYBODY TALKS ABOUT IT...   $200   
3                      THE COMPANY LINE   $200   
4                   EPITAPHS & TRIBUTES   $200   
...                                 ...    ...   
216925                   RIDDLE ME THIS  $2000   
216926                        "T" BIRDS  $2000   
216927           AUTHORS IN THEIR YOUTH  $2000   
216928                       QUOTATIONS  $2000   
216929                   HISTORIC NAMES   None   

                                                                                                                      Question  \
0                             For the last 8 years of his life, Galileo was under house arrest for espousing this man's theory   
1                  No. 2: 1912 Olympian; football star at Carlisle Indian School; 6 MLB seasons with the Reds, Giants & Braves   
2                                     The city of Yuma in this state has a record average of 4,055 hours of sunshine each year   
3                                         In 1963, live on "The Art Linkletter Show", this company served its billionth burger   
4                     Signer of the Dec. of Indep., framer of the Constitution of Mass., second President of the United States   
...                                                                                                                        ...   
216925                                              This Puccini opera turns on the solution to 3 riddles posed by the heroine   
216926                 In North America this term is properly applied to only 4 species that are crested, including the tufted   
216927               In Penny Lane, where this "Hellraiser" grew up, the barber shaves another customer--then flays him alive!   
216928  From Ft. Sill, Okla. he made the plea, Arizona is my land, my home, my father's land, to which I now ask to... return"   
216929                 A silent movie title includes the last name of this 18th c. statesman & favorite of Catherine the Great   

                                Answer  
0                           Copernicus  
1                           Jim Thorpe  
2                              Arizona  
3                           McDonald's  
4                           John Adams  
...                                ...  
216925                        Turandot  
216926                      a titmouse  
216927                    Clive Barker  
216928                        Geronimo  
216929  Grigori Alexandrovich Potemkin  

[216930 rows x 7 columns]

I write a function that filters the dataset for questions that contain all of the words in a list of words. For example, when the list `["King", "England"]` is passed to the function, I should obtain a dataframe of 49 rows, with each row having both strings `"King"` and `"England"` somewhere in the question.

In [4]:
query = ["King", "England"]
question_number = -1 # This will indicate the row the question came from
question_rows = [] # List of row numbers from which to build this dataframe

for question in all_data.Question: # Check each question
    FoundFlag = True # Create a flag that switches if any word is not found in the question
    question_number += 1 # Track the question we are inspecting
    for term in query:
        if term not in question: # If one term is not in the question, ignore this question
            FoundFlag = False # Activate switch
    if FoundFlag == True:
        question_rows.append(question_number) # add the locations of valid questions to the list

print(len(all_data.iloc[question_rows]))
all_data.iloc[question_rows]

49


Show Number    Air Date             Round  \
4953           3003  1997-09-24  Double Jeopardy!   
14912          2832  1996-12-17         Jeopardy!   
21511          4650  2004-11-19         Jeopardy!   
23810          4862  2005-11-01         Jeopardy!   
27555          1799  1992-05-28  Double Jeopardy!   
33294          5589  2008-12-18         Jeopardy!   
41148          5925  2010-05-21  Double Jeopardy!   
41357          2751  1996-07-15         Jeopardy!   
43122          3937  2001-10-16  Double Jeopardy!   
47814          4365  2003-07-18  Double Jeopardy!   
49994          5499  2008-07-03         Jeopardy!   
51115          3574  2000-03-02         Jeopardy!   
51565          4264  2003-02-27  Double Jeopardy!   
56600          3533  2000-01-05  Double Jeopardy!   
57516          4299  2003-04-17  Double Jeopardy!   
58949          6250  2011-11-18  Double Jeopardy!   
71808          5270  2007-07-06         Jeopardy!   
74235          2757  1996-09-03         Jeopardy!   
77195          3656  2000-06-26         Jeopardy!   
79457          4230  2003-01-10  Double Jeopardy!   
80113          5827  2010-01-05         Jeopardy!   
84639          1558  1991-05-15         Jeopardy!   
86412          1837  1992-09-08  Double Jeopardy!   
89710          5390  2008-02-01  Double Jeopardy!   
111926         1331  1990-05-21         Jeopardy!   
118809         1905  1992-12-11         Jeopardy!   
123299         4656  2004-11-29         Jeopardy!   
123671         5718  2009-06-17         Jeopardy!   
125822         4042  2002-03-12  Double Jeopardy!   
127484         1325  1990-05-11  Double Jeopardy!   
127758         6021  2010-11-15         Jeopardy!   
129106         4906  2006-01-02  Double Jeopardy!   
130208         5222  2007-05-01         Jeopardy!   
137818         3368  1999-04-07         Jeopardy!   
143434         1882  1992-11-10  Double Jeopardy!   
150880         4790  2005-06-03  Double Jeopardy!   
165110         4255  2003-02-14         Jeopardy!   
165197         6272  2011-12-20  Double Jeopardy!   
170007         5455  2008-05-02  Double Jeopardy!   
174055         2143  1993-12-22  Double Jeopardy!   
175259         3055  1997-12-05  Double Jeopardy!   
183462         3099  1998-02-05   Final Jeopardy!   
186239         5339  2007-11-22  Double Jeopardy!   
187651         5264  2007-06-28         Jeopardy!   
191579         5453  2008-04-30         Jeopardy!   
192265         6292  2012-01-17  Double Jeopardy!   
192966         6175  2011-06-17  Double Jeopardy!   
194516         6114  2011-03-24  Double Jeopardy!   
200369         2937  1997-05-13         Jeopardy!   

                            Category   Value  \
4953                    "PH"UN WORDS    $200   
14912                  WORLD HISTORY    $100   
21511                     THE "O.C."   $1000   
23810                  NAME THE YEAR    $400   
27555               HISTORIC IN-LAWS    $600   
33294            THE BAYEUX TAPESTRY    $600   
41148                       TREATIES   $1600   
41357                 HISTORIC NAMES    $400   
43122                     OH, HENRY!    $400   
47814                        POTIONS   $2000   
49994                    REIGN DROPS   $1000   
51115             ALL THINGS BRITISH    $200   
51565             MUSIC APPRECIATION    $800   
56600                 FRENCH HISTORY    $600   
57516               THE "UN"CATEGORY   $2000   
58949            ASSUME THE POSITION   $2000   
71808        WHERE THERE'S A WILLIAM   $1000   
74235                  WORLD HISTORY    $200   
77195                        JUNE 26    $400   
79457                     CONQUERORS    $800   
80113                      "M" & "M"   $1000   
84639              FILMS OF THE '90S    $200   
86412                        ROYALTY    $400   
89710                 "CAN"NED WORDS   $1600   
111926                FILE UNDER "M"  $1,000   
118809                KINGS & QUEENS    $300   
123299                        CHAIRS    $400   
123671       

This test has worked, so I now turn this into a function.

In [5]:
def questions_by_query(query): # Define this function

    question_number = -1 # Reset the question number
    question_rows = [] # Empty list of row numbers
    
    for question in all_data.Question: # Check each question
        FoundFlag = True # Reset flag for each question
        question_number += 1 # Track the question we are inspecting    
        for term in query:
            if term.upper() not in question.upper(): # Remove case-sensitivity
                FoundFlag = False
        if FoundFlag == True:
            question_rows.append(question_number)

    return all_data.iloc[question_rows]

query = ["Game"]
questions_by_query(query)

Show Number    Air Date             Round                 Category  \
67             5957  2010-07-06         Jeopardy!        RHYMES WITH SMART   
72             5957  2010-07-06         Jeopardy!             LET'S BOUNCE   
174            3751  2000-12-18   Final Jeopardy!           SPORTS LEGENDS   
218            3673  2000-07-19  Double Jeopardy!  INTERNATIONAL SPORTSMEN   
433            6037  2010-12-07         Jeopardy!             LET'S HIT IT   
...             ...         ...               ...                      ...   
216067         4364  2003-07-17  Double Jeopardy!        PULLING UP STAKES   
216125         6207  2011-09-20  Double Jeopardy!        ISN'T IT "GRAND"?   
216167         1983  1993-03-31         Jeopardy!                   SPORTS   
216281         5236  2007-05-21         Jeopardy!      ALL WE ARE "SAY"ING   
216709         6044  2010-12-16         Jeopardy!           LANGUAGE CRAFT   

       Value  \
67      $400   
72      $600   
174     None   
218     $600   
433     $600   
...      ...   
216067  $800   
216125  $800   
216167  $400   
216281  $600   
216709  $800   

                                                                                                                                                                                   Question  \
67                                                                                                                                       Small, slender missile thrown at a board in a game   
72                                                                                              In this kid's game, you bounce a small rubber ball while picking up 6-pronged metal objects   
174                                                                   If Joe DiMaggio's hitting streak had gone one more game in 1941, this company would have given him a $10,000 contract   
218                                                                                                            Vladimir Samsonov is touted as Europe's only hope against China in this game   
433                                                                                                          Aaron Fechter invented this carnival game where you hit a mammal with a mallet   
...                                                                                                                                                                                     ...   
216067                                                                                                          A ringer on the stake in this game scores 3 points; 50 points wins the game   
216125                                                                                                 Titles in this mega-selling video game series include "Vice City" & "Chinatown Wars"   
216167                                                                                         In 1975 this college football bowl game moved from Tulane Stadium to the Louisiana Superdome   
216281                                                                                            This popular children's game bearing a man's name tests your ability to follow directions   
216709  ("Daylight" playing) Matt & Kim reworked their song "Daylight" for the gibberish language used in <a href="http://www.j-archive.com/media/2010-12-16_J_21.wmv">this</a> game series   

                                 Answer  
67                               a dart  
72                                jacks  
174     H.J. Heinz (Heinz 57 Varieties)  
218                           Ping-pong  
433                        Whack-A-Mole  
...                                 ...  
216067                       horseshoes  
216125                 Grand Theft Auto  
216167                   the Sugar Bowl  
216281                       Simon Says  
216709                         The Sims  

[2083 rows x 7 columns]

Here, we can see that the function only checks that the group of letters making up the word are in the question. This includes not just the word `"game"`, but `"games"` too. I attempt to fix this by checking for additional letters on either side of this string of letters.

In [6]:
import string
string.ascii_uppercase

'ABCDEFGHIJKLMNOPQRSTUVWXYZ'

In [7]:
def questions_by_strict_query(query): # Define this function

    question_number = -1 # Reset the question number
    question_rows = [] # Empty list of row numbers
    
    for question in all_data.Question: # Check each question
        FoundFlag = True # Reset flag for each question
        question_number += 1 # Track the question we are inspecting    
        for term in query:
            BadTermVersions = [] # Make an array of letters on either side of the search term
            for x in string.ascii_uppercase:
                BadTermVersions.append(x+term.upper())
                BadTermVersions.append(term.upper()+x)

            if term.upper() not in question.upper(): # Remove case-sensitivity
                FoundFlag = False
            else: # Raw term is found in the question
                BadInstances = 0
                Instances = question.upper().count(term.upper()) # number of times the raw term appears in the question
                for BadTerm in BadTermVersions:
                    if BadTerm in question.upper(): # if the term is part of a longer string
                        BadInstances += question.upper().count(BadTerm)
                        # Double counting when the string has extra letter on both sides: XXabcXX
                    BadInstances -= question.upper().count(term.upper()+x+term.upper())
                if BadInstances == Instances: # if all instances are bad
                    FoundFlag = False
        if FoundFlag == True:
            question_rows.append(question_number)

    return all_data.iloc[question_rows]

query = ["Games"]
questions_by_strict_query(query)

Show Number    Air Date             Round  \
918            3834  2001-04-12  Double Jeopardy!   
1528           5392  2008-02-05  Double Jeopardy!   
2318           5630  2009-02-13  Double Jeopardy!   
2598           4220  2002-12-27  Double Jeopardy!   
2737           5979  2010-09-16  Double Jeopardy!   
...             ...         ...               ...   
213102         4156  2002-09-30  Double Jeopardy!   
213525          351  1986-01-13         Jeopardy!   
214035         6011  2010-11-01  Double Jeopardy!   
215175         5167  2007-02-13  Double Jeopardy!   
216052         4364  2003-07-17         Jeopardy!   

                            Category  Value  \
918     THE MELBOURNE OLYMPICS, 1956   $200   
1528               FUNNY FOR NOTHIN'   $800   
2318               CANADIAN FOOTBALL  $2000   
2598                THE EMPEROR NERO   $400   
2737                   WORLD LEADERS  $2000   
...                              ...    ...   
213102                  LA TRIVIA-TA  $2000   
213525                         CARDS   $100   
214035               IT'S HYPHENATED   $400   
215175                PLENTY OF TIME   $400   
216052           THE AMERICAN LEAGUE  $1000   

                                                                                                                         Question  \
918           As an 8-year-old in 1956, she sang with a group welcoming Queen Elizabeth II; later she sang to open the 2000 games   
1528    A writer, on the U.S. soccer team's 4 total shots in 3 games: "Four shots?" This Laker "takes that many during a timeout"   
2318                              The Alouettes play their home games at Molson Stadium on the campus of this Montreal university   
2598                                                   In the year prior to his death, Nero participated in these games in Greece   
2737                             It's prime minister Mr. Tuila'Epa, won a silver medal in archery at the 2007 South Pacific Games   
...                                                                                                                           ...   
213102                                                    The eagle that represented the 1984 Summer Olympic Games was named this   
213525                                              Bridge term for a set of three games or what a hot rod lays when it takes off   
214035                                        With 7, Nolan Ryan holds the record for having pitched the most of these rare games   
215175      This sport's games have no set time limit; the longest pro game was 8 hours, 25 minutes between Pawtucket & Rochester   
216052                      On Sept. 27, 1999 this team played the last of its 6,783 games at "The Corner" of Michigan & Trumbull   

                    Answer  
918     Olivia Newton-John  
1528           Kobe Bryant  
2318                McGill  
2598         Olympic Games  
2737                 Samoa  
...                    ...  
213102                 Sam  
213525              rubber  
214035         a no-hitter  
215175            baseball  
216052      Detroit Tigers  

[448 rows x 7 columns]

I format the values in the Value column to be floats, so that further calculations may be performed more easily on those values.

In [8]:
all_data

Show Number    Air Date             Round  \
0              4680  2004-12-31         Jeopardy!   
1              4680  2004-12-31         Jeopardy!   
2              4680  2004-12-31         Jeopardy!   
3              4680  2004-12-31         Jeopardy!   
4              4680  2004-12-31         Jeopardy!   
...             ...         ...               ...   
216925         4999  2006-05-11  Double Jeopardy!   
216926         4999  2006-05-11  Double Jeopardy!   
216927         4999  2006-05-11  Double Jeopardy!   
216928         4999  2006-05-11  Double Jeopardy!   
216929         4999  2006-05-11   Final Jeopardy!   

                               Category  Value  \
0                               HISTORY   $200   
1       ESPN's TOP 10 ALL-TIME ATHLETES   $200   
2           EVERYBODY TALKS ABOUT IT...   $200   
3                      THE COMPANY LINE   $200   
4                   EPITAPHS & TRIBUTES   $200   
...                                 ...    ...   
216925                   RIDDLE ME THIS  $2000   
216926                        "T" BIRDS  $2000   
216927           AUTHORS IN THEIR YOUTH  $2000   
216928                       QUOTATIONS  $2000   
216929                   HISTORIC NAMES   None   

                                                                                                                      Question  \
0                             For the last 8 years of his life, Galileo was under house arrest for espousing this man's theory   
1                  No. 2: 1912 Olympian; football star at Carlisle Indian School; 6 MLB seasons with the Reds, Giants & Braves   
2                                     The city of Yuma in this state has a record average of 4,055 hours of sunshine each year   
3                                         In 1963, live on "The Art Linkletter Show", this company served its billionth burger   
4                     Signer of the Dec. of Indep., framer of the Constitution of Mass., second President of the United States   
...                                                                                                                        ...   
216925                                              This Puccini opera turns on the solution to 3 riddles posed by the heroine   
216926                 In North America this term is properly applied to only 4 species that are crested, including the tufted   
216927               In Penny Lane, where this "Hellraiser" grew up, the barber shaves another customer--then flays him alive!   
216928  From Ft. Sill, Okla. he made the plea, Arizona is my land, my home, my father's land, to which I now ask to... return"   
216929                 A silent movie title includes the last name of this 18th c. statesman & favorite of Catherine the Great   

                                Answer  
0                           Copernicus  
1                           Jim Thorpe  
2                              Arizona  
3                           McDonald's  
4                           John Adams  
...                                ...  
216925                        Turandot  
216926                      a titmouse  
216927                    Clive Barker  
216928                        Geronimo  
216929  Grigori Alexandrovich Potemkin  

[216930 rows x 7 columns]

There is a value of 'None' at row ID 216,929. I will replace all values of None in this column to a format that matches the other values, before filtering.

In [9]:
all_data['Value'] = all_data['Value'].replace('None', '$0')
all_data

Show Number    Air Date             Round  \
0              4680  2004-12-31         Jeopardy!   
1              4680  2004-12-31         Jeopardy!   
2              4680  2004-12-31         Jeopardy!   
3              4680  2004-12-31         Jeopardy!   
4              4680  2004-12-31         Jeopardy!   
...             ...         ...               ...   
216925         4999  2006-05-11  Double Jeopardy!   
216926         4999  2006-05-11  Double Jeopardy!   
216927         4999  2006-05-11  Double Jeopardy!   
216928         4999  2006-05-11  Double Jeopardy!   
216929         4999  2006-05-11   Final Jeopardy!   

                               Category  Value  \
0                               HISTORY   $200   
1       ESPN's TOP 10 ALL-TIME ATHLETES   $200   
2           EVERYBODY TALKS ABOUT IT...   $200   
3                      THE COMPANY LINE   $200   
4                   EPITAPHS & TRIBUTES   $200   
...                                 ...    ...   
216925                   RIDDLE ME THIS  $2000   
216926                        "T" BIRDS  $2000   
216927           AUTHORS IN THEIR YOUTH  $2000   
216928                       QUOTATIONS  $2000   
216929                   HISTORIC NAMES     $0   

                                                                                                                      Question  \
0                             For the last 8 years of his life, Galileo was under house arrest for espousing this man's theory   
1                  No. 2: 1912 Olympian; football star at Carlisle Indian School; 6 MLB seasons with the Reds, Giants & Braves   
2                                     The city of Yuma in this state has a record average of 4,055 hours of sunshine each year   
3                                         In 1963, live on "The Art Linkletter Show", this company served its billionth burger   
4                     Signer of the Dec. of Indep., framer of the Constitution of Mass., second President of the United States   
...                                                                                                                        ...   
216925                                              This Puccini opera turns on the solution to 3 riddles posed by the heroine   
216926                 In North America this term is properly applied to only 4 species that are crested, including the tufted   
216927               In Penny Lane, where this "Hellraiser" grew up, the barber shaves another customer--then flays him alive!   
216928  From Ft. Sill, Okla. he made the plea, Arizona is my land, my home, my father's land, to which I now ask to... return"   
216929                 A silent movie title includes the last name of this 18th c. statesman & favorite of Catherine the Great   

                                Answer  
0                           Copernicus  
1                           Jim Thorpe  
2                              Arizona  
3                           McDonald's  
4                           John Adams  
...                                ...  
216925                        Turandot  
216926                      a titmouse  
216927                    Clive Barker  
216928                        Geronimo  
216929  Grigori Alexandrovich Potemkin  

[216930 rows x 7 columns]

In [10]:
def currency_to_float(value):
    return value[1:].replace(',', '')

all_data['Value'] = all_data['Value'].apply(currency_to_float).apply(pd.to_numeric)
all_data

Show Number    Air Date             Round  \
0              4680  2004-12-31         Jeopardy!   
1              4680  2004-12-31         Jeopardy!   
2              4680  2004-12-31         Jeopardy!   
3              4680  2004-12-31         Jeopardy!   
4              4680  2004-12-31         Jeopardy!   
...             ...         ...               ...   
216925         4999  2006-05-11  Double Jeopardy!   
216926         4999  2006-05-11  Double Jeopardy!   
216927         4999  2006-05-11  Double Jeopardy!   
216928         4999  2006-05-11  Double Jeopardy!   
216929         4999  2006-05-11   Final Jeopardy!   

                               Category  Value  \
0                               HISTORY    200   
1       ESPN's TOP 10 ALL-TIME ATHLETES    200   
2           EVERYBODY TALKS ABOUT IT...    200   
3                      THE COMPANY LINE    200   
4                   EPITAPHS & TRIBUTES    200   
...                                 ...    ...   
216925                   RIDDLE ME THIS   2000   
216926                        "T" BIRDS   2000   
216927           AUTHORS IN THEIR YOUTH   2000   
216928                       QUOTATIONS   2000   
216929                   HISTORIC NAMES      0   

                                                                                                                      Question  \
0                             For the last 8 years of his life, Galileo was under house arrest for espousing this man's theory   
1                  No. 2: 1912 Olympian; football star at Carlisle Indian School; 6 MLB seasons with the Reds, Giants & Braves   
2                                     The city of Yuma in this state has a record average of 4,055 hours of sunshine each year   
3                                         In 1963, live on "The Art Linkletter Show", this company served its billionth burger   
4                     Signer of the Dec. of Indep., framer of the Constitution of Mass., second President of the United States   
...                                                                                                                        ...   
216925                                              This Puccini opera turns on the solution to 3 riddles posed by the heroine   
216926                 In North America this term is properly applied to only 4 species that are crested, including the tufted   
216927               In Penny Lane, where this "Hellraiser" grew up, the barber shaves another customer--then flays him alive!   
216928  From Ft. Sill, Okla. he made the plea, Arizona is my land, my home, my father's land, to which I now ask to... return"   
216929                 A silent movie title includes the last name of this 18th c. statesman & favorite of Catherine the Great   

                                Answer  
0                           Copernicus  
1                           Jim Thorpe  
2                              Arizona  
3                           McDonald's  
4                           John Adams  
...                                ...  
216925                        Turandot  
216926                      a titmouse  
216927                    Clive Barker  
216928                        Geronimo  
216929  Grigori Alexandrovich Potemkin  

[216930 rows x 7 columns]

I create a function to count the number of unique answers.

In [11]:
def count_answers(data):

    UniqueAnswers = {}                       # dictionary of answer and the number of appearances 
    for answer in data['Answer']:
        if answer not in UniqueAnswers:      # if new answer
            UniqueAnswers[answer] = 1        # add to dictionary
        else:
            UniqueAnswers[answer] += 1
    return UniqueAnswers

def query_to_answers(query):                    # processing function so I can just input the query
    filtered = questions_by_strict_query(query) # get the dataset filtered by the query
    answers = count_answers(filtered)           # count the number of answers in the filtered data
    return filtered, answers

In [12]:
query = ['bread']
filtered_data, answer_counts = query_to_answers(query) # array type needs to be passed to function

I can display the most common answers.

In [13]:
import numpy as np

In [14]:
answer_counts

keys = list(answer_counts.keys())
values = list(answer_counts.values())
sorted_dict = {keys[i]: values[i] for i in np.argsort(values)[::-1]} # argsort returns a sorted index array
sorted_dict

print("For the query of:\n")
for q in query:
    print(q)
print('')
for k,v in sorted_dict.items():
    print("An answer of", k, "appears", v, "times.")

For the query of:

bread

An answer of French toast appears 5 times.
An answer of Passover appears 5 times.
An answer of pita appears 5 times.
An answer of Wonder Bread appears 4 times.
An answer of baguette appears 4 times.
An answer of rye appears 4 times.
An answer of focaccia appears 3 times.
An answer of croutons appears 3 times.
An answer of a baguette appears 3 times.
An answer of Sourdough appears 3 times.
An answer of mold appears 3 times.
An answer of Jean Valjean appears 2 times.
An answer of Zwieback appears 2 times.
An answer of yeast appears 2 times.
An answer of English muffin appears 2 times.
An answer of pumpernickel appears 2 times.
An answer of caraway seeds appears 2 times.
An answer of stuffing appears 2 times.
An answer of Elvis Presley appears 2 times.
An answer of matzo appears 2 times.
An answer of zwieback appears 2 times.
An answer of Smorgasbord appears 2 times.
An answer of Old Mother Hubbard appears 2 times.
An answer of scones appears 2 times.
An answer o

I can sort the filtered dataset by value, with the highest-scoring questions at the top.

In [15]:
filtered_data.sort_values(by=['Value'], ascending=False)

Show Number    Air Date             Round                  Category  \
176999         2998  1997-09-17  Double Jeopardy!                     TREES   
122510         4615  2004-10-01  Double Jeopardy!             CHART TOPPERS   
75793          6077  2011-02-01  Double Jeopardy!                       EAT   
99486          5633  2009-02-18  Double Jeopardy!  CAN I BUY YOU A SANDWCH?   
119624         5445  2008-04-18  Double Jeopardy!                      EGGS   
...             ...         ...               ...                       ...   
89719          3023  1997-10-22         Jeopardy!     INTERNATIONAL CUISINE   
155220         3947  2001-10-30         Jeopardy!                     BREAD   
156721           63  1984-12-05         Jeopardy!           FOREIGN CUISINE   
71843          2921  1997-04-21         Jeopardy!                     BREAD   
15694          6260  2011-12-02   Final Jeopardy!              WORD ORIGINS   

        Value  \
176999   2000   
122510   2000   
75793    2000   
99486    2000   
119624   2000   
...       ...   
89719     100   
155220    100   
156721    100   
71843     100   
15694       0   

                                                                                                                   Question  \
176999                                  Saint-John's Bread is another name for this tree that yields a chocolate substitute   
122510  He wrote Bread's only No. 1 hit, "Make It With You", as well as the title track to the 1977 film "The Goodbye Girl"   
75793                         You can make your panini with this Italian bread whose name comes from the Latin for "hearth"   
99486                   The name of this hot pressed sandwich made with a variety of fillings is Italian for "little bread"   
119624           For Passover, this Jewish dish is sort of like French toast but uses unleavened bread dipped in beaten egg   
...                                                                                                                     ...   
89719                                        It's traditional to cut this religious symbol into the top of Irish soda bread   
155220     It's the term for a type of French bread baked into a long, cylindrical loaf with a crisp crust & a chewy inside   
156721                                                                               The Middle East's popular pocket bread   
71843                                Mince cloves of this into butter to prepare a yummy bread found in Italian restaurants   
15694                                       This word for a friend comes from the Latin for "with whom you would eat bread"   

             Answer  
176999        Carob  
122510  David Gates  
75793      focaccia  
99486        panini  
119624   matzo brei  
...             ...  
89719         Cross  
155220     baguette  
156721       a pita  
71843        Garlic  
15694     companion  

[265 rows x 7 columns]

I can count how many questions there are for each discrete amount of points.

In [16]:
filtered_data['Value'].value_counts().sort_index(ascending=False)

2000     9
1600    12
1500     1
1200    13
1000    27
800     37
600     21
500      9
400     58
300     20
200     47
100     10
0        1
Name: Value, dtype: int64

I can also count how many questions were in each type of round.

In [17]:
filtered_data['Round'].value_counts()

Jeopardy!           170
Double Jeopardy!     94
Final Jeopardy!       1
Name: Round, dtype: int64

I can sort the dataset by points and round type, in an order with the most useful questions at the top.

In [18]:
# Define the order in which to sort the strings
filtered_data['Round'] = pd.Categorical(filtered_data['Round'], ["Final Jeopardy!", "Double Jeopardy!", "Jeopardy!"])

C:\Users\k_cha\AppData\Local\Temp\ipykernel_18108\454882251.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  filtered_data['Round'] = pd.Categorical(filtered_data['Round'], ["Final Jeopardy!", "Double Jeopardy!", "Jeopardy!"])


In [19]:
filtered_data.sort_values(by=['Round', 'Value'], ascending=[True, False])

Show Number    Air Date             Round                  Category  \
15694          6260  2011-12-02   Final Jeopardy!              WORD ORIGINS   
75793          6077  2011-02-01  Double Jeopardy!                       EAT   
99486          5633  2009-02-18  Double Jeopardy!  CAN I BUY YOU A SANDWCH?   
119624         5445  2008-04-18  Double Jeopardy!                      EGGS   
122510         4615  2004-10-01  Double Jeopardy!             CHART TOPPERS   
...             ...         ...               ...                       ...   
89719          3023  1997-10-22         Jeopardy!     INTERNATIONAL CUISINE   
112859         2813  1996-11-20         Jeopardy!               WORLD FACTS   
155220         3947  2001-10-30         Jeopardy!                     BREAD   
156721           63  1984-12-05         Jeopardy!           FOREIGN CUISINE   
193219         3057  1997-12-09         Jeopardy!           A LOAF OF BREAD   

        Value  \
15694       0   
75793    2000   
99486    2000   
119624   2000   
122510   2000   
...       ...   
89719     100   
112859    100   
155220    100   
156721    100   
193219    100   

                                                                                                                   Question  \
15694                                       This word for a friend comes from the Latin for "with whom you would eat bread"   
75793                         You can make your panini with this Italian bread whose name comes from the Latin for "hearth"   
99486                   The name of this hot pressed sandwich made with a variety of fillings is Italian for "little bread"   
119624           For Passover, this Jewish dish is sort of like French toast but uses unleavened bread dipped in beaten egg   
122510  He wrote Bread's only No. 1 hit, "Make It With You", as well as the title track to the 1977 film "The Goodbye Girl"   
...                                                                                                                     ...   
89719                                        It's traditional to cut this religious symbol into the top of Irish soda bread   
112859                                                 Romanians enjoy eating mamaliga, a bread or mush made from this meal   
155220     It's the term for a type of French bread baked into a long, cylindrical loaf with a crisp crust & a chewy inside   
156721                                                                               The Middle East's popular pocket bread   
193219                                            This bread made with a "starter" got its start among old-time prospectors   

             Answer  
15694     companion  
75793      focaccia  
99486        panini  
119624   matzo brei  
122510  David Gates  
...             ...  
89719         Cross  
112859     Cornmeal  
155220     baguette  
156721       a pita  
193219    Sourdough  

[265 rows x 7 columns]

This shows the best questions to study at the top, first ordering by Round, with priority given to Final Jeopardy, before the value of the question is sorted, with the highest-earners at the top.

#### I now build a quizzing system.

Filter the dataset to only include shows that have questions from each round.

In [54]:
FullShows = []
for i in range(1, np.max(all_data['Show Number']+1)):
    if i not in all_data['Show Number'].values:
        print("Show", i, "is not in the database.")
    else:
        data = all_data.loc[all_data['Show Number'] == i]
        if len(data.loc[data['Round'] == 'Jeopardy!']) < 1:
            print('No questions for show', i)
        if len(data.loc[data['Round'] == 'Double Jeopardy!']) < 1:
            print('No Double Jeopardy! questions for show', i)
        if len(data.loc[data['Round'] == 'Final Jeopardy!']) < 1:
            print('No Final Jeopardy! questions for show', i)
        else:
            FullShows.append(i)

Show 6 is not in the database.
Show 7 is not in the database.
Show 8 is not in the database.
Show 15 is not in the database.
Show 16 is not in the database.
Show 17 is not in the database.
Show 18 is not in the database.
Show 19 is not in the database.
Show 20 is not in the database.
Show 21 is not in the database.
Show 22 is not in the database.
Show 23 is not in the database.
Show 24 is not in the database.
Show 25 is not in the database.
Show 26 is not in the database.
Show 27 is not in the database.
Show 28 is not in the database.
Show 29 is not in the database.
Show 30 is not in the database.
Show 31 is not in the database.
Show 32 is not in the database.
Show 33 is not in the database.
Show 34 is not in the database.
Show 35 is not in the database.
Show 36 is not in the database.
Show 37 is not in the database.
Show 38 is not in the database.
Show 39 is not in the database.
Show 40 is not in the database.
Show 41 is not in the database.
Show 42 is not in the database.
Show 43 is 

Show 506 is not in the database.
Show 507 is not in the database.
Show 508 is not in the database.
Show 509 is not in the database.
Show 510 is not in the database.
Show 511 is not in the database.
Show 512 is not in the database.
Show 513 is not in the database.
Show 514 is not in the database.
Show 515 is not in the database.
Show 516 is not in the database.
Show 517 is not in the database.
Show 518 is not in the database.
Show 519 is not in the database.
Show 520 is not in the database.
Show 521 is not in the database.
Show 522 is not in the database.
Show 524 is not in the database.
Show 525 is not in the database.
Show 526 is not in the database.
Show 527 is not in the database.
Show 528 is not in the database.
Show 529 is not in the database.
Show 530 is not in the database.
Show 531 is not in the database.
Show 532 is not in the database.
Show 534 is not in the database.
Show 535 is not in the database.
Show 536 is not in the database.
Show 537 is not in the database.
Show 538 i

Show 1049 is not in the database.
Show 1050 is not in the database.
Show 1051 is not in the database.
Show 1052 is not in the database.
Show 1053 is not in the database.
Show 1054 is not in the database.
Show 1055 is not in the database.
Show 1056 is not in the database.
Show 1057 is not in the database.
Show 1058 is not in the database.
Show 1059 is not in the database.
Show 1060 is not in the database.
Show 1061 is not in the database.
Show 1062 is not in the database.
Show 1063 is not in the database.
Show 1064 is not in the database.
Show 1065 is not in the database.
Show 1066 is not in the database.
Show 1067 is not in the database.
Show 1068 is not in the database.
Show 1069 is not in the database.
Show 1070 is not in the database.
Show 1071 is not in the database.
Show 1072 is not in the database.
Show 1073 is not in the database.
Show 1074 is not in the database.
Show 1075 is not in the database.
Show 1076 is not in the database.
Show 1077 is not in the database.
Show 1078 is n

Show 1783 is not in the database.
Show 1784 is not in the database.
Show 1785 is not in the database.
Show 1786 is not in the database.
Show 1787 is not in the database.
Show 1788 is not in the database.
Show 1789 is not in the database.
Show 1796 is not in the database.
Show 1797 is not in the database.
Show 1798 is not in the database.
Show 1801 is not in the database.
Show 1802 is not in the database.
Show 1803 is not in the database.
Show 1804 is not in the database.
Show 1805 is not in the database.
Show 1806 is not in the database.
Show 1807 is not in the database.
Show 1808 is not in the database.
Show 1809 is not in the database.
Show 1810 is not in the database.
Show 1811 is not in the database.
Show 1812 is not in the database.
Show 1813 is not in the database.
Show 1814 is not in the database.
Show 1815 is not in the database.
Show 1816 is not in the database.
Show 1817 is not in the database.
Show 1818 is not in the database.
Show 1819 is not in the database.
Show 1820 is n

Show 2087 is not in the database.
Show 2088 is not in the database.
Show 2089 is not in the database.
Show 2090 is not in the database.
Show 2091 is not in the database.
Show 2092 is not in the database.
Show 2093 is not in the database.
Show 2094 is not in the database.
Show 2095 is not in the database.
Show 2096 is not in the database.
Show 2097 is not in the database.
Show 2098 is not in the database.
Show 2099 is not in the database.
Show 2100 is not in the database.
Show 2101 is not in the database.
Show 2102 is not in the database.
Show 2103 is not in the database.
Show 2104 is not in the database.
Show 2105 is not in the database.
Show 2106 is not in the database.
Show 2107 is not in the database.
Show 2108 is not in the database.
Show 2109 is not in the database.
Show 2110 is not in the database.
Show 2111 is not in the database.
Show 2112 is not in the database.
Show 2113 is not in the database.
Show 2114 is not in the database.
Show 2115 is not in the database.
Show 2131 is n

Show 2516 is not in the database.
Show 2517 is not in the database.
Show 2518 is not in the database.
Show 2519 is not in the database.
Show 2520 is not in the database.
Show 2521 is not in the database.
Show 2522 is not in the database.
Show 2523 is not in the database.
Show 2524 is not in the database.
Show 2525 is not in the database.
Show 2526 is not in the database.
Show 2527 is not in the database.
Show 2528 is not in the database.
Show 2529 is not in the database.
Show 2530 is not in the database.
Show 2531 is not in the database.
Show 2532 is not in the database.
Show 2533 is not in the database.
Show 2534 is not in the database.
Show 2535 is not in the database.
Show 2536 is not in the database.
Show 2537 is not in the database.
Show 2538 is not in the database.
Show 2539 is not in the database.
Show 2540 is not in the database.
Show 2542 is not in the database.
Show 2544 is not in the database.
Show 2545 is not in the database.
Show 2546 is not in the database.
Show 2547 is n

No Double Jeopardy! questions for show 3022
No Final Jeopardy! questions for show 3022
Show 3074 is not in the database.
Show 3176 is not in the database.
Show 3238 is not in the database.
Show 3300 is not in the database.
Show 3406 is not in the database.
Show 3536 is not in the database.
Show 3572 is not in the database.
Show 3594 is not in the database.
Show 3607 is not in the database.
Show 3636 is not in the database.
No Double Jeopardy! questions for show 3656
No Final Jeopardy! questions for show 3656
Show 3866 is not in the database.
Show 3869 is not in the database.
Show 3894 is not in the database.
Show 3895 is not in the database.
Show 3923 is not in the database.
Show 3924 is not in the database.
Show 3925 is not in the database.
Show 3953 is not in the database.
Show 3954 is not in the database.
Show 3955 is not in the database.
Show 3956 is not in the database.
Show 3957 is not in the database.
Show 3958 is not in the database.
Show 3959 is not in the database.
Show 3960 

No Double Jeopardy! questions for show 6086
No Final Jeopardy! questions for show 6086
No questions for show 6087
Show 6094 is not in the database.
Show 6299 is not in the database.
No Double Jeopardy! questions for show 6300
No Final Jeopardy! questions for show 6300


In [55]:
print('There are', len(FullShows), 'shows that have a complete set of questions.')

There are 3631 shows that have a complete set of questions.


In [22]:
import random as rand

In [ ]:
def quiz(FullShows, Difficulty):
    # Set the difficulty of the opponent
    if Difficulty == 'Easier':
        DifficultyArray = [True, True, False, False, False, False, False, False, False, False]
    elif Difficulty == 'Harder':
        DifficultyArray = [True, True, True, True, True, True, True, True, False, False]
    elif Difficulty == 'Random':
        DifficultyArray = [True, False]
    ComputerPoints = 0
    PlayerPoints = 0
    player_control = 'User'
    # Pick a random show
    show_number = FullShows[rand.randint(0,len(FullShows)-1)]
    # Filter the dataset to get the questions from each round of this show
    show_questions = all_data.loc[all_data['Show Number'] == show_number]
    # Group the dataset by round
    Round1 = show_questions.loc[show_questions['Round'] == "Jeopardy!"].reset_index()
    Round2 = show_questions.loc[show_questions['Round'] == "Double Jeopardy!"].reset_index()
    Round3 = show_questions.loc[show_questions['Round'] == "Final Jeopardy!"].reset_index()
    # Iterate by round - the first two have similar gameplay
    for r in range(0, 2):
        if r == 0: Round=Round1
        if r == 1: Round=Round2
        # Ask every question
        round_in_play = True
        while round_in_play == True:
            print(Round[['Category', 'Value']])
            if player_control == 'User':
                questionChoice = input('Select your question by entering the number on the left:   ')
            else:
                questionChoice = rand.randint(0, len(Round)-1)
            current_question = Round.iloc[[questionChoice]]
            Category = current_question.Category.item()
            Question = current_question.Question.item()
            CorrectAnswer = current_question.Answer.item()
            Value = current_question.Value.item()
            # Who chose this question?
            if player_control == 'User':
                print('You have chosen', Category, 'for', Value, 'points!')
            else:
                print('The Computer has chosen', Category, 'for', Value, 'points!')
            print(Question)
            # Answer the question
            if player_control == 'User':
                user_answer = input('Who/what is...')
                # User answers correctly
                if user_answer == CorrectAnswer:
                    PlayerPoints += Value
                    print('You have won', Value,'points!')
                # User is wrong - play goes to the computer
                else:
                    computer_answer = rand.choice(DifficultyArray)
                    # Computer answers correctly - control goes to the Computer
                    if computer_answer == True:
                        ComputerPoints += Value
                        player_control = 'Computer'
                        print('The Computer won', Value, 'points! The correct answer was',
                              CorrectAnswer + '.')
                    # Nobody answers correctly
                    else:
                        print('Nobody won. The correct answer was', CorrectAnswer)
            # Control starts with the computer
            else:
                computer_answer = rand.choice(DifficultyArray)
                # Computer answers correctly
                if computer_answer == True:
                    ComputerPoints += Value
                    print('The Computer won', Value,'points! The correct answer was', CorrectAnswer, +'.')
                # Computer is wrong
                else:
                    user_answer = input('Who/what is...')
                    # User answers correctly - control goes to the user
                    if user_answer == CorrectAnswer:
                        player_control = 'User'
                        PlayerPoints += Value
                        print('You have won', Value,'points!')
                    # Nobody answers correctly
                    else:
                        print('Nobody won. The correct answer was', CorrectAnswer)
            # Points summary
            print('The Computer has', str(ComputerPoints) + '. You have', str(PlayerPoints) +'.')
            # Question is removed from the dataset
            Round = Round.drop([Round.index[int(questionChoice)]]).reset_index(drop=True)
            if len(Round) < 1:
                round_in_play = False
                print('At the end of Round', str(r+1) + ', you have', str(PlayerPoints),
                      'and The Computer has', str(ComputerPoints) + 'points!')
    # Now in Round 3
    print('The final category is:   ', Round3.iloc[[0]].Category.item())
    user_wager = input('How much would you like to wager on knowing the answer to this question?   ')
    computer_wager = rand.randint(1, ComputerPoints)
    print ('The Computer has bet', computer_wager, 'points!')
    print('The question is:')
    print(Round3.iloc[[0]].Question.item())
    user_answer = input('Who/what is...')
    computer_answer = rand.choice(DifficultyArray)
    # User's final score
    if user_answer == Round3.iloc[[0]].Answer.item():
        PlayerPoints += user_wager
        print('You have won your wager and your final score is', PlayerPoints, 'points!')
    else:
        PlayerPoints -= user_wager
        print('You have lost your wager and your final score is', PlayerPoints, 'points!')
    # The Computer's final score
    if computer_answer == True:
        ComputerPoints += computer_wager
        print('The Computer won its wager and its final score is', ComputerPoints, 'points!')
    else:
        ComputerPoints -= computer_wager
        print('The Computer lost its wager and your final score is', ComputerPoints, 'points!')
    print('At the end of the game, you have', str(PlayerPoints), 'and The Computer has', str(ComputerPoints) + 'points!')
    if ComputerPoints > PlayerPoints:
        print('The Computer beat you! Train some more!')
    elif ComputerPoints == PlayerPoints:
        print('You drew! Ready for a rematch?')
    else:
        print('You won! Congratulations!')
quiz(FullShows, 'Easier')